packages-exp/auth-exp/src/platform_browser/messagechannel/receiver.ts

```diff
@@ -0,0 +1,161 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  ReceiverHandler,
+  EventType,
+  ReceiverResponse,
+  SenderMessageEvent,
+  Status,
+  SenderRequest
+} from './index';
+import { _allSettled } from './promise';
+
+/**
+ * Interface class for receiving messages.
+ *
+ * @internal
+ */
+export class Receiver {
+  private static readonly receivers: Receiver[] = [];
+  private readonly boundEventHandler: EventListener;
+
+  private readonly handlersMap: {
+    // Typescript doesn't have existential types :(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    [eventType: string]: Set<ReceiverHandler<any, any>>;
+  } = {};
+
+  constructor(private readonly eventTarget: EventTarget) {
+    this.boundEventHandler = this.handleEvent.bind(this);
+  }
+
+  /**
+   * Obtain an instance of a Receiver for a given event target, if none exists it will be created.
+   *
+   * @param eventTarget - An event target (such as window or self) through which the underlying
+   * messages will be received.
+   */
+  static _getInstance(eventTarget: EventTarget): Receiver {
+    // The results are stored in an array since objects can't be keys for other
+    // objects. In addition, setting a unique property on an event target as a
```

Comment:
What about a `Map<key, value>`? Keys can be objects

Reply:
```
Type 'EventTarget' cannot be used as an index type.
```

Reply:
Ah yeah because it's a DOM interface. That's fine then